### ISSUE TITLE
Intermittent blocked stream

### ISSUE THREAD

**Issue by davedoesdev** [fails-components/webtransport#78]
Hi-
In the test below, there are two bidirectional streams between client and server. 
The server reads from the first stream continuously and so we should see displayed constant output from stream1. 
It only reads from stream2 once so we should see only one output from stream2.
However, in about a third of the runs, stream1 comes to a halt after about a second and we see no output from it again.
On the normal runs, output is as expected.
Is there a way of obtaining trace/debug output from quiche to see why stream1 is being blocked?
Thanks

```js
import { createReadStream } from 'fs';
import { Writable } from 'stream';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { readFile } from 'fs/promises';
import { X509Certificate } from 'crypto';
import { Http3Server, WebTransport } from '@fails-components/webtransport';

const certs_dir = join(dirname(fileURLToPath(import.meta.url)), '..', 'certs');
const cert = await readFile(join(certs_dir, 'server.crt'));

const server = new Http3Server({
    port: 8080,
    host: '127.0.0.1',
    secret: 'testsecret',
    cert,
    privKey: await readFile(join(certs_dir, 'server.key'))
});

(async () => {
    try {
        const session_stream = server.sessionStream('/test');
        const session_reader = session_stream.getReader();

        const session = await session_reader.read();
        if (session.done) {
            return console.log('Server finished');
        }
        console.log('got a new session');
        await session.value.ready;
        console.log('server session is ready');

        const bidi_reader = session.value.incomingBidirectionalStreams.getReader();
        const bidi_stream1 = await bidi_reader.read();
        if (bidi_stream1.done) {
            console.log('bidi reader done');
            return;
        }
        if (bidi_stream1.value) {
            console.log('got bidi stream');
        }

        const bidi_stream2 = await bidi_reader.read();
        if (bidi_stream2.done) {
            console.log('bidi reader done');
            return;
        }
        if (bidi_stream2.value) {
            console.log('got bidi stream');
        }

        (async () => {
            const reader = bidi_stream1.value.readable.getReader();
            while (true) { // eslint-disable-line no-constant-condition
                const data = await reader.read();
                if (data.done) {
                    return;
                }
                console.log('stream1', data.value.length);
            }
        })();

        (async () => {
            const reader = bidi_stream2.value.readable.getReader();
            const data = await reader.read();
            if (data.done) {
                return;
            }
            console.log('stream2', data.value.length);
        })();
    } catch (ex) {
        console.error('Server error', ex);
    }
})();

server.startServer();

await new Promise(resolve => setTimeout(resolve, 2000));

const client = new WebTransport('https://127.0.0.1:8080/test', {
    serverCertificateHashes: [{
        algorithm: 'sha-256',
        value: Buffer.from(
            new X509Certificate(cert).fingerprint256.split(':').map(
                el => parseInt(el, 16)))
    }]
});
client.closed
    .then(() => {
        console.log('Connection closed');
    })
    .catch(err => {
        console.error('Connection errored', err);
    });
console.log('Waiting for client to be ready');
await client.ready;
console.log('Client ready');

const stream1 = await client.createBidirectionalStream();
const stream2 = await client.createBidirectionalStream();
console.log('Client created streams');

createReadStream('/dev/urandom').pipe(Writable.fromWeb(stream1.writable));
createReadStream('/dev/urandom').pipe(Writable.fromWeb(stream2.writable));
```

**Comment by martenrichter** [comment]
Ok, I tried to reproduce it, which was the most complicated, since I am still on node 16 LTS, which does not have the `fromWeb` method, so I had to move to node 18 for the test.
First I thought you might have found a bug, but I have many intermittent streams already in my project.
Anyway, the solution is simple. You are using the same file `/dev/urandom` for producing the input, which can be blocked, by the second stream, since you are not reading anymore. This is the intended behavior of the streaming stuff.
If you use `/dev/null` for one of the streams the problem does not occur anymore.

Regarding debugging, if it is inside the library, I am afraid, that so far I have turned off the debug output by quiche, since it is massive and I have no idea how to integrate it with node in a useful way. 
But it will not help you with blocking streams. I use most times a transform stream with debugging output in between to see at which point a stream is blocking.

**Comment by davedoesdev** [comment]
`fs.createReadStream` [opens a new file descriptor](https://github.com/nodejs/node/blob/main/lib/internal/fs/streams.js#L69) so not reading the input can't be the cause.

For example, this program doesn't block even though it stops reading from stream 2:

```js
import { createReadStream } from 'fs';
import { Writable } from 'stream';

createReadStream('/dev/urandom').pipe(new Writable({
    write(chunk, endoding, cb) {
        console.log('stream1', chunk.length);
        cb();
    }
}));

createReadStream('/dev/urandom').pipe(new Writable({
    write(chunk, endoding, cb) {
        console.log('stream2', chunk.length);
    }
}));
```

The reason using `/dev/null` works for one of the streams is it produces no data so it's like having a single stream.

**Comment by martenrichter** [comment]
Ok, you are right that `/dev/null` was the wrong choice, can you test it with `/dev/zero` instead? I really want to rule out, that nothing outside the lib is causing the blocking and the webstream interface of node is also still experimental. (And if already tore down the testing environment for node 18, that I used in the morning)

Since the long weekend is over, I can only look into this next weekend.
For diagnosis, you may want to uncomment inside the `node_modules` directory in the package's webtransport.js` the `console.log` inside the various `callback` methods. In this way, you can see, what is flowing out from the `c++` part. Also, the methods passed to the readable and writable of the `Http3WTStream` are a good place to start peak into the incoming data.

**Comment by martenrichter** [comment]
And I ran into some issues with the node.js event loop:
https://github.com/nodejs/node/discussions/43882
Where the event loop was stalled.
However, I switched to node API, and the issues were gone.

**Comment by davedoesdev** [comment]
Two things I found so far:

- Also repros if I use `/dev/zero` for both streams (about one in three runs).
- The data isn't written because `QuicStream::CanWriteNewData` returns `false` on the first stream (65536 bytes buffered, threashold is 8192).
- I can repro every time _and_ the first stream gets _no data at all_ by modifying the test as below. This should make it easier to debug. You'll see in this case I'm using `/dev/urandom` for the first stream and `/dev/zero` for the second. This suggests to me it's a timing issue and `/dev/zero` is producing data faster.

```js
import { createReadStream } from 'fs';
import { Writable, Transform } from 'stream';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { readFile } from 'fs/promises';
import { X509Certificate } from 'crypto';
import { Http3Server, WebTransport } from '@fails-components/webtransport';

const certs_dir = join(dirname(fileURLToPath(import.meta.url)), '..', 'certs');
const cert = await readFile(join(certs_dir, 'server.crt'));

const server = new Http3Server({
    port: 8080,
    host: '127.0.0.1',
    secret: 'testsecret',
    cert,
    privKey: await readFile(join(certs_dir, 'server.key'))
});

(async () => {
    try {
        const session_stream = server.sessionStream('/test');
        const session_reader = session_stream.getReader();

        const session = await session_reader.read();
        if (session.done) {
            return console.log('Server finished');
        }
        console.log('got a new session');
        await session.value.ready;
        console.log('server session is ready');

        const bidi_reader = session.value.incomingBidirectionalStreams.getReader();
        const bidi_stream1 = await bidi_reader.read();
        if (bidi_stream1.done) {
            console.log('bidi reader done');
            return;
        }
        if (bidi_stream1.value) {
            console.log('got bidi stream');
        }

        const bidi_stream2 = await bidi_reader.read();
        if (bidi_stream2.done) {
            console.log('bidi reader done');
            return;
        }
        if (bidi_stream2.value) {
            console.log('got bidi stream');
        }

        (async () => {
            const reader = bidi_stream1.value.readable.getReader();
            while (true) { // eslint-disable-line no-constant-condition
                const data = await reader.read();
                if (data.done) {
                    return;
                }
                console.log('stream1', data.value.length);
            }
        })();

        (async () => {
            const reader = bidi_stream2.value.readable.getReader();
            const data = await reader.read();
            if (data.done) {
                return;
            }
            console.log('stream2', data.value.length);
        })();
    } catch (ex) {
        console.error('Server error', ex);
    }
})();

server.startServer();

await new Promise(resolve => setTimeout(resolve, 2000));

const client = new WebTransport('https://127.0.0.1:8080/test', {
    serverCertificateHashes: [{
        algorithm: 'sha-256',
        value: Buffer.from(
            new X509Certificate(cert).fingerprint256.split(':').map(
                el => parseInt(el, 16)))
    }]
});
client.closed
    .then(() => {
        console.log('Connection closed');
    })
    .catch(err => {
        console.error('Connection errored', err);
    });
console.log('Waiting for client to be ready');
await client.ready;
console.log('Client ready');

const stream1 = await client.createBidirectionalStream();
const stream2 = await client.createBidirectionalStream();
console.log('Client created streams');

const writer1 = stream1.writable.getWriter();
const writer2 = stream2.writable.getWriter();

stream1.wup = 'stream1';
stream2.wup = 'stream2';

createReadStream('/dev/urandom').pipe(new Writable({
    async write(chunk, encoding, cb) {
        try {
            await writer1.write(chunk);
            cb();
        } catch (ex) {
            cb(ex);
        }
    }
}));
createReadStream('/dev/zero').pipe(new Writable({
    async write(chunk, encoding, cb) {
        try {
            await writer2.write(chunk);
            cb();
        } catch (ex) {
            cb(ex);
        }
    }
}));
```

**Comment by davedoesdev** [comment]
Maybe I'm assuming (wrongly) that WebTransport applies backpressure on each stream individually rather than on the session.

I was assuming that just because one stream in a session is backed up with data, it wouldn't mean other streams in the session couldn't transfer data.

**Comment by martenrichter** [comment]
Thanks, these are good pointers. And so far all audio and video stuff had very low data rates so I never crossed the 65536 bytes barrier, which explains, why it was not triggered before.
Backpressure should be handled by every stream individually, as far as I understand the protocol, but it may be some other intermediate buffer, that runs full.
I will take a closer look at the weekend and I hope, it is fixable on my side and not on quiche.

**Comment by davedoesdev** [comment]
OK I think I've got to the bottom of it.
The `/dev/urandom` ends up being marked blocked in `QuicStream::WriteBufferedData`.
This is because `PriorityWriteScheduler::ShouldYield` says it should yield to the `/dev/zero` stream which is at the same priority and isn't being read from. As it's faster to generate data, it has already filled the session's buffer.

Since the `/dev/zero` stream is only read from once, the session can never write again, so `QuicSession::OnCanWrite` never calls `write_blocked_streams_.PopFront` to unblock the `/dev/urandom` stream.

The related discussions for WebTransport seem to be:

https://github.com/w3c/webtransport/issues/33
https://github.com/w3c/webtransport/issues/62

where it's discussed exposing adding QUIC's priority setting to WebTransport streams. However, the issue is still open.

It would seem to me that QUIC sessions don't push the data into the receiver to make sure they can't be blocked but instead rely on applications to specify priority.

**Comment by davedoesdev** [comment]
I can confirm that. If I modify `WebTransportStream`, `WebTransportStreamAdapter` and `Http3WTStream` to add `SetPriority`, and then set the priority of the `/dev/urandom` stream to 2 (higher) and the priority of the `/dev/zero` stream to 3 (lower) then the issue doesn't occur.

**Comment by martenrichter** [comment]
Well, I am not completely sure, if this is not only a symptom. I need to check if it is not in fact a problem in the stream visitor, which does not try to resend if it was blocked before. 
Also, I just read quickly through the issues, and for me it seems, that the issues talk about datagrams and not the streams, but I may be mistaken.

**Comment by martenrichter** [comment]
Ok, confirmed that the stream visitor is working properly. So it seems to be, that your diagnosis is correct.
The question is, what to do about it?
I am a bit reluctant to add the priority setting as long as it is not part of the spec. (Especially since it does not prevent a client on the browser side to do the flooding).
I would say, that the application should add a transform stream, that limits the data rate and that an unlimited data rate stream is not a realistic test case.

**Comment by davedoesdev** [comment]
I too am wondering if this is a symptom. With further reading, I gather that backpressure should be applied per stream as you said, _as well as_ per session.
I wonder if the buffers in the streams need to be big enough - maybe quiche doesn't provide its own buffers. I can dig into it more when I get time.

**Comment by davedoesdev** [comment]
OK I've got to the bottom of this.

If I add this code to `http3server.c`:

```js
  const uint32_t kInitialSessionFlowControlWindow = 1 * 1024 * 1024;  // 1 MB
  const uint32_t kInitialStreamFlowControlWindow = 64 * 1024;         // 64 KB
  if (sconfig.GetInitialStreamFlowControlWindowToSend() ==
      kDefaultFlowControlSendWindow) {
    sconfig.SetInitialStreamFlowControlWindowToSend(
        kInitialStreamFlowControlWindow);
  }
  if (sconfig.GetInitialSessionFlowControlWindowToSend() ==
      kDefaultFlowControlSendWindow) {
    sconfig.SetInitialSessionFlowControlWindowToSend(
        kInitialSessionFlowControlWindow);
  }
```

then the `/dev/urandom` stream doesn't block. The reason is that by default the streams and the session both have 16kb windows. So when the `/dev/zero` stream fills up its window, the window of the session is also filled up. When the session windows is larger than the stream windows then there is still space for the non-blocked stream to send its data.

The above code is from `QuicServer::Initialize` and I put it before

```js
      server_ = std::make_unique<Http3Server>(eventloop, host, port, std::move(proofsource), secret.c_str(), sconfig);
```

in `http3server.cc`.

**Comment by martenrichter** [comment]
Thank you for going to the bottom of this.
I must have forgotten these lines or thought that they may be addressed later if problems occur and that it would save some memory.
I have added a commit to include the suggested changes:
https://github.com/fails-components/webtransport/pull/79/commits/0a928a1039bea9d4772d99495ced200778755539

Anyway, do you have an opinion, on if these settings need to be settable from the javascript side? Or should most people be able to live with the defaults?

I will probably create an release at the weekend with these changes and the latest changes from libquiche.

**Comment by davedoesdev** [comment]
Unfortunately I think yes it should be configurable from JS,, especially useful if expecting a lot of streams.

**Comment by martenrichter** [comment]
Ok, I have added logic for this:
https://github.com/fails-components/webtransport/pull/80/commits/1834c56b61598108c881eb3894f73e6560b6459b
probably it is also a good idea for some applications to lower the value to save memory.

**Comment by davedoesdev** [comment]
Thanks, I've tested it and it works!
